Feature/add total member

Kenny-Gin1 opened the pull request:
Issue: https://github.com/DXgovernance/dxdao-contracts/issues/129

- Added a `totalHolder` count for our `ERC20SnapshotRep.sol` contract
- Added a mock for testing `internal` functions and simplifying `mint` functions
- Added tests

Was trying to figure out how to test internal functions and simplify minting. The best practice seemed to just make an inheritance contract overriding functions and exposing internal functions. Let me know if this is the way we should do it, and if there are eventual other issues with this approach.

rossneilson commented on contracts/utils/ERC20/ERC20SnapshotRep.sol:
```diff
@@ -4,11 +4,17 @@ pragma solidity ^0.8.8;
 import "@openzeppelin/contracts-upgradeable/token/ERC20/extensions/ERC20SnapshotUpgradeable.sol";
 import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
 import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/math/SafeMathUpgradeable.sol";
```
We shouldnt need safemath for 0.8 right?

AugustoL replied:
I think we dont, but we are still using the openzeppelin libs.

AugustoL commented on contracts/test/ERC20SnapshotRepMock.sol:
```diff
@@ -0,0 +1,30 @@
+contract ERC20SnapshotRepMock is ERC20SnapshotUpgradeable, ERC20SnapshotRep {
+    constructor() {}
+
+    function mint(address to, uint256 amount) public override {
```
This funcion is not needed, the mint are burned are defined in the ERC20SnapshotRep for only owner

AugustoL commented on contracts/test/ERC20SnapshotRepMock.sol:
```diff
@@ -0,0 +1,30 @@
+        return removeHolders(account);
+    }
+
+    function _beforeTokenTransfer(
```
looks like thsi fixes a wront use of _beforeTokenTransfer in ERC20SnapshotRep, where the _beforeTokenTransfer uses onlyOwner.

Just remove the 
```
    function _beforeTokenTransfer(
        address from,
        address to,
        uint256 amount
    ) internal virtual override onlyOwner {}
    ``` in the ERC20SnapshotRep and you should be able to remove this _beforeTokenTransfer and everything should work.

Kenny-Gin1 commented:
@rossneilson @AugustoL Thanks for your feedback I've now fixed it in regards to that.

- Changed from plural to singular regarding `addHolder` and `removeHolder` functions
- Removed the `mint` and `_beforeTokenTransfer` functions
- Changed the tests to better reflect usage
- Removed `SafeMath` library since Solidity checks for overflows by default in version 0.8 <

AugustoL commented:
Also, does all tests run succesfully? the test actions failed https://github.com/DXgovernance/dxdao-contracts/runs/6162734754?check_suite_focus=true

Kenny-Gin1 commented:
> Also, does all tests run succesfully? the test actions failed https://github.com/DXgovernance/dxdao-contracts/runs/6162734754?check_suite_focus=true

Yes! I will look at it together today with @MiltonTulli before merging. Because I think it's in regards to calling the `mint` in `ERC20SnapshotRep`which has the `_snapshot()` function in it.

AugustoL commented:
one tests in failing in the CI

AugustoL reviewed (changes requested):
@Kenny-Gin1 try merging with develop, fix tests and commit

Kenny-Gin1 commented:
@AugustoL @MiltonTulli realised that removing this part:     

```
function _beforeTokenTransfer(
        address from,
        address to,
        uint256 amount
    ) internal virtual override onlyOwner {}
```

from the `ERC20SnapshotRep` made one of @MiltonTulli tests fail. Should it still be removed? Are we testing it wrongly for some reason?

Kenny-Gin1 commented:
Alright, all tests are working ✅  @rossneilson @AugustoL

AugustoL commented on package.json:
```diff
@@ -102,8 +102,7 @@
     "*.{ts,tsx,js,jsx,json,md,sol}": [
       "yarn format",
       "yarn solidity-linter",
-      "git add",
```
remove git add too please.

After commit if you have changes left you can add them with `git add -p` and `git commit --amend`

MiltonTulli commented on test/utils/ERC20/ERC20SnapshotRep.js:
```diff
@@ -0,0 +1,57 @@
+      from: accounts[0],
+    });
+    await ERC20SnapshotRep.initialize("DXdao", "DXD");
+    await ERC20SnapshotRep._addHolder(SOME_ADDRESS, { from: accounts[0] });
```
@Kenny-Gin1 I think that you can omit `, {from : accounts[0]}` in all contract function executions. By default will execute with the owner (L.14). 

```
await ERC20SnapshotRep._addHolder(SOME_ADDRESS);
await ERC20SnapshotRep._removeHolder(SOME_ADDRESS);
await ERC20SnapshotRep.mint(SOME_ADDRESS, 100);
await ERC20SnapshotRep.burn(SOME_ADDRESS, 50);
```
